lib/spack/spack/spec.py

```diff
@@ -4358,79 +4338,73 @@ def format(self, format_string=DEFAULT_FORMAT, **kwargs):
         literal ``\`` character.
 
         Args:
-            format_string (str): string containing the format to be expanded
+            format_string: string containing the format to be expanded
 
         Keyword Args:
-            color (bool): True if returned string is colored
-            transform (dict): maps full-string formats to a callable \
-                that accepts a string and returns another one
-
+            color: True if returned string is colored, False if not,
+                None for auto color (default is False)
         """
         ensure_modern_format_string(format_string)
-        color = kwargs.get("color", False)
-        transform = kwargs.get("transform", {})
-
-        out = io.StringIO()
-
-        def write(s, c=None):
-            f = clr.cescape(s)
-            if c is not None:
-                f = COLOR_FORMATS[c] + f + "@."
-            clr.cwrite(f, stream=out, color=color)
-
-        def write_attribute(spec, attribute, color):
-            attribute = attribute.lower()
 
-            sig = ""
-            if attribute.startswith(("@", "%", "/")):
-                # color sigils that are inside braces
-                sig = attribute[0]
-                attribute = attribute[1:]
-            elif attribute.startswith("arch="):
-                sig = " arch="  # include space as separator
-                attribute = attribute[5:]
+        # TODO: This preserves a bug from the original version of this function that's needed
+        # TODO: by format_path(). Get rid of it eventually; plain '\' should just format as '\'
+        if format_string == "\\":
+            return ""
```

Comment:
@scheibelp @johnwparent: this seems to be needed for `format_path`, the only place I see that formats a raw `\` string. It would be nice if it were handled there instead of in `Spec.format()`.  It's a question of whether you think a single backslash should format as `""` or a single backslash. I think the latter makes more sense than this special case.

Reply:
`format_path` shouldn't be passing `\` into this method when it operates as a path separator (e.g. on Windows): it splits on path separators and formats the components individually.

Is there a test you see failing because of this?

> It's a question of whether you think a single backslash should format as "" (old Spec.format() did this) or a single backslash

I'd prefer `\` to `""`, but I'd prefer raising an exception to doing either: technically `\` is allowed in Linux paths, but it's discouraged and I don't see why we should seek out a headache (if `format_path` is letting any of those through, I'd say it's a bug, or a user is formatting `\\` in a path on Linux, which we might as well tell them not to do). 

Even though `format` is used for things other than paths, I'm not sure what allowing `\` achieves for those cases.

Reply:
> Is there a test you see failing because of this?

If you remove this check and add an assert like so:

```diff
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 520dd73f84..81bb1e33bc 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -4351,8 +4351,8 @@ def format(self, format_string: str = DEFAULT_FORMAT, color: Optional[bool] = Fa

         # TODO: This preserves a bug from the original version of this function that's needed
         # TODO: by format_path(). Get rid of it eventually; plain '\' should just format as '\'
-        if format_string == "\\":
-            return ""
+        #        if format_string == "\\":
+        #            return ""

         def safe_color(sigil: str, string: str, color_fmt: Optional[str]):
             # avoid colorizing if there is no color or the string is empty
@@ -4513,6 +4513,9 @@ def format_path(
         else:
             output_path_components = []
             input_path_components = list(format_string_as_path.parts)
+
+        assert input_path_components[0] != "\\"
+
         output_path_components += [
             fs.polite_filename(self.format(x)) for x in input_path_components
         ]
```

then run:

```console
$ spack unit-test --showlocals -vv 'lib/spack/spack/test/spec_semantics.py::test_spec_format_path_windows'
```

You get:

```console
lib/spack/spack/test/spec_semantics.py:1128: in _check_spec_format_path
    formatted = spec.format_path(format_str, _path_ctor=path_ctor)
        expected   = 'installroot\\git-test\\git.foo_bar'
        format_str = '/installroot/{name}/{version}'
        path_ctor  = <class 'pathlib.PureWindowsPath'>
        spec       = git-test@git.foo/bar
        spec_str   = 'git-test@git.foo/bar'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = git-test@git.foo/bar, format_string = '/installroot/{name}/{version}'
_path_ctor = <class 'pathlib.PureWindowsPath'>

    def format_path(
        # self, format_string: str, _path_ctor: Optional[pathlib.PurePath] = None
        self,
        format_string: str,
        _path_ctor: Optional[Callable[[Any], pathlib.PurePath]] = None,
    ) -> str:
        """Given a `format_string` that is intended as a path, generate a string
        like from `Spec.format`, but eliminate extra path separators introduced by
        formatting of Spec properties.

        Path separators explicitly added to the string are preserved, so for example
        "{name}/{version}" would generate a directory based on the Spec's name, and
        a subdirectory based on its version; this function guarantees though that
        the resulting string would only have two directories (i.e. that if under
        normal circumstances that `str(Spec.version)` would contain a path
        separator, it would not in this case).
        """
        format_component_with_sep = r"\{[^}]*[/\\][^}]*}"
        if re.search(format_component_with_sep, format_string):
            raise SpecFormatPathError(
                f"Invalid path format string: cannot contain {{/...}}\n\t{format_string}"
            )

        path_ctor = _path_ctor or pathlib.PurePath
        format_string_as_path = path_ctor(format_string)
        if format_string_as_path.is_absolute():
            output_path_components = [format_string_as_path.parts[0]]
            input_path_components = list(format_string_as_path.parts[1:])
        else:
            output_path_components = []
            input_path_components = list(format_string_as_path.parts)

>       assert input_path_components[0] != "\\"
E       AssertionError

_path_ctor = <class 'pathlib.PureWindowsPath'>
format_component_with_sep = '\\{[^}]*[/\\\\][^}]*}'
format_string = '/installroot/{name}/{version}'
format_string_as_path = PureWindowsPath('/installroot/{name}/{version}')
input_path_components = ['\\', 'installroot', '{name}', '{version}']
output_path_components = []
path_ctor  = <class 'pathlib.PureWindowsPath'>
self       = git-test@git.foo/bar
```

Reply:
That is specifically for the 

`r"/installroot/{name}/{version}"`

test case on windows, which is a nonsense path. As long as it formats to something we expect, the output can be anything (doesn't have to be a valid path)

In other words, it's fine to set `expected` to `_\installroot\git-test\git.foo_bar` in this case (and remove these lines):

(`['\\', 'installroot', '{name}', '{version}']` are the input path components, and `polite_filename` makes sure `\\` is transformed into `_` so it has no significance as a directory)

Does that seem agreeable?

Reply:
So personally I think it would be more consistent if `format_path` noticed this:

```
input_path_components = ['\\', 'installroot', '{name}', '{version}']
```

And just removed the leading slash if it is relevant on windows. If what you say is true (`format_path` should not be passing a single `\` to format()) then the behavior seems way better to me, and the expected result is not dependent on what is essentially a corner case for `format()`, i.e., an escape character by itself.  I can add that to this PR if you think it makes sense.

Reply:
This does make me wonder, though, whether `format()` shouldn't just behave like Python and use `{{` and `}}` as its escape characters instead of `\{` and `\}`.  Would that eliminate the need to handle path formatting specially?  In a lot of ways I wonder why we need `format_path` and whether anyone's actually escaping their `{` or `}` characters in spack format strings.

Reply:
ok, after looking at the code a bit, I decided to do two things:
1. raise an error in `Spec.format()` if it's passed an unterminated escape, i.e. `'\'` or `'foobar\'`
2. preserve any leading `'/'` as `'\'`, even if it's inconsequential. This is what `PureWindowsPath` does, so I think it's the most reasonable.

We could probably eliminate the need for `format_path` by using double-brace escapes but that can be some other PR.

Reply:
@scheibelp does this work for you?

Reply:
> We could probably eliminate the need for format_path by using double-brace escapes but that can be some other PR.

`format_path` isn't about what goes in: changing from `{` to `{{` would not fix what it is made for. `format_path` makes sure that Spec attributes like `Spec.version` don't generate characters that either (a) look like path separators or (b) look like other potentially problematic characters (e.g. `=` in makefiles).

At least, I don't see how switching to `{{` would address that, so I could use some more explanation to understand what you mean (not here necessarily, since you didn't make that change here, but whenever you intend to do it).

> raise an error in Spec.format() if it's passed an unterminated escape, i.e. '\' or 'foobar\'

Looking at https://github.com/spack/spack/pull/43712/commits/14c8f70358d2b74ba9870e9eed133430eb6ab553 I'm a bit worried about the complexity of SPEC_FORMAT_RE (and e.g. have trouble seeing what happens if an unterminated escape appears somewhere besides the end of the string), but the overall desired behavior sounds good to me.

> preserve any leading '/' as '\', even if it's inconsequential. This is what PureWindowsPath does, so I think it's the most reasonable.

Users wouldn't want the result regardless, so it all seems fine to me (including what you chose), although the way you achieved this seems somewhat roundabout:

```
        output_path_components += [
            fs.polite_filename(self.format(x)) for x in input_path_components
            fs.polite_filename(self.format(part)) if part != "\\" else part
            for part in input_path_components
        ]
```

this logic appears to consider cases of `\\` that are not just at the beginning: if we only intend to deal with leading `\\` in this way, I think this code should be explicit about it.

I'd recommend a comment there about how a '\\` can even come up, given that the string was just fed to something that "should" remove these.

Reply:
> `format_path` isn't about what goes in: changing from `{` to `{{` would not fix what it is made for. `format_path` makes sure that Spec attributes like `Spec.version` don't generate characters that either (a) look like path separators or (b) look like other potentially problematic characters (e.g. `=` in makefiles).

Ok, sorry I wasn't clear. It would eliminate the need to split the path by separator, but I think it has its own set of special cases, and it's not critical. I don't think it is a particularly needed change.

>  I'm a bit worried about the complexity of SPEC_FORMAT_RE

I don't think we can avoid this -- that regex is the whole reason that this is fast.

> e.g. have trouble seeing what happens if an unterminated escape appears somewhere besides the end of the string

That doesn't happen. If it occurs in the string option 1 consumes it and the character after `\` is output. The only way option 1 doesn't match is if there is a `\` at the end of the string; otherwise there is always something after the `\`.

> the overall desired behavior sounds good to me.

that's good!

> although the way you achieved this seems somewhat roundabout:
> 
> ```
>         output_path_components += [
>             fs.polite_filename(self.format(x)) for x in input_path_components
>             fs.polite_filename(self.format(part)) if part != "\\" else part
>             for part in input_path_components
>         ]
> ```
> 
> this logic appears to consider cases of `\\` that are not just at the beginning: if we only intend to deal with leading `\\` in this way, I think this code should be explicit about it.
> 
> I'd recommend a comment there about how a '\\` can even come up, given that the string was just fed to something that "should" remove these.

I investigated this further. As you say, the *only* place you can have a `\\` *on Windows* is in fact the beginning of the string. I was thinking you could get it somewhere else, but almost any other place I put backslashes they're ignored or treated as path separators. I think there are three cases on Windows:

* Initial `C:\` (or other letter) is absolute and results from a path like `C:\\foo/bar/baz`
* Initial `\\server\c\` is a remote server path and volume. It's interpreted as absolute and results from a path like `\\server\c\foo\bar\baz`
* Initial `\` happens when you pass a path that begins with `\` or `/`

The first two are handled by the existing logic; the last one is not.  However, the last one can occur on Posix.  In fact, any path component can end with `\\` on Posix, which will break `format()`:

```
>>> Path(r"\/foo\/bar/baz").parts
('\\', 'foo\\', 'bar', 'baz')
```

So the handling actually needs to be different here. Any path component can legitimately end with an unterminated `\\`.  Yes it's unlikely but it is allowed.

I think based on this, the reasonable thing to do for a final `\\` is *not* to raise but to just have `format()` leave the `\\` alone.  That would solve this in one place and would simplify `SPEC_FORMAT_RE` as well as the logic in `format_path`.  Agree?

Reply:
> Ok, sorry I wasn't clear. It would eliminate the need to split the path by separator, but I think it has its own set of special cases, and it's not critical. I don't think it is a particularly needed change.

I still don't 100% follow (but again, that doesn't worry me here): `format_path` not only splits on path separators, but also calls `polite_filename` on each component.

> I think based on this, the reasonable thing to do for a final \\ is not to raise but to just have format() leave the \\ alone (i.e. just preserve it). That would solve this in one place and would simplify SPEC_FORMAT_RE as well as the logic in format_path. I can do that and add a comment. Agree?

I think this means you are talking about modifying the Windows test like described in https://github.com/spack/spack/pull/43712#discussion_r1571872941 - if I understood that correctly, then I agree

Reply:
Per discussion with Peter we decided to:
* Not raise for an unterminated `\` in format strings (i.e., leave it untouched)
* On windows, pass an initial `\` through as we would for an absolute path, so that the result of `format_path()` is consistent with `PureWindowsPath` on Windows.
* On posix, treat an initial `\` as a path component (as it must be on posix) and format it.r

This simplifies the ragex and the logic in `format_path`